File: Path Sum/sol.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def hasPathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: bool
        """
        return self.traverse_tree(root, targetSum)
        
    def traverse_tree(self, node, current_sum): 
        if node is None: 
            return False 
        
        new_sum = current_sum - node.val
        if(node.left is None and node.right is None and  new_sum == 0):
            return True 
        
        return self.traverse_tree(node.left, new_sum) or self.traverse_tree(node.right, new_sum)

File: Path Sum/sol.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        
        return traverseTree(root, targetSum);
    }
    
    private Boolean traverseTree(TreeNode node, int currentSum) {
        if(node == null) {
            return false; 
        }
        
        int newSum = currentSum - node.val; 
        if(node.left == null && node.right == null && newSum == 0) {
            return true;
        }
        
        return traverseTree(node.left, newSum) || traverseTree(node.right, newSum);
    }
}